ow_regolith/src/RegolithSpawner.cpp

```diff
@@ -0,0 +1,295 @@
+    ROS_ERROR("Connection to service %s has been lost", 
+      srv.getService().c_str()
+    );
+    return false;
+  }
+  if (!srv.call(msg)) {
+    ROS_ERROR("Failed to call service %s", srv.getService().c_str());
+    return false;
+  } 
+  return true;
+}
+
+template <class T>
+static bool createRosServiceClient(unique_ptr<ros::NodeHandle> &nh, 
+  string service_path, ros::ServiceClient &out_srv, bool persistent = false)
+{
+  // maximum time we will wait for the service to exist
+  constexpr auto SERVICE_TIMEOUT = 5.0f; // seconds
+  out_srv = nh->serviceClient<T>(service_path, persistent);
+  if (!out_srv.waitForExistence(ros::Duration(SERVICE_TIMEOUT))) {
+    ROS_ERROR("Timed out waiting for service %s to advertise", 
+      service_path.c_str()
+    );
+    return false;
+  } 
+  return true;
+}
+
+RegolithSpawner::RegolithSpawner(ros::NodeHandle* nh)
+  : m_volume_displaced(0.0), m_node_handle(nh) 
+{
+  // get node parameters, which mirror class data member
+  if (!m_node_handle->getParam("spawn_volume_threshold", m_spawn_threshold))
+    ROS_ERROR("Regolith node requires the spawn_volume_threshold parameter.");
+  if (!m_node_handle->getParam("regolith_model_uri", m_model_uri))
+    ROS_ERROR("Regolith node requires the regolith_model_uri paramter.");
+}
+
+bool RegolithSpawner::initialize() 
+{
+  // set the maximum scoop inclination that the psuedo force can counteract
+  // to be 45 degrees (excluding friction)
+  // NOTE: do not use a value that makes cosine zero!
+  constexpr auto MAX_SCOOP_INCLINATION_DEG = 70.0f; // degrees
+  constexpr auto MAX_SCOOP_INCLINATION_RAD 
+    = MAX_SCOOP_INCLINATION_DEG * acos(-1) / 180.0f; // radians
+  constexpr auto PSUEDO_FORCE_WEIGHT_FACTOR 
+    = 1.0f / cos(MAX_SCOOP_INCLINATION_RAD);
+
+  // load SDF model
+  if (!getSdfFromUri(m_model_uri, m_model_sdf)) {
+    ROS_ERROR("Failed to load SDF for regolith model");
+    return false;
+  }
+  // parse as SDF object to query properties about the model
+  auto sdf = parseSdf(m_model_sdf);
+  float model_mass;
+  if (!getModelLinkMass(sdf, model_mass) ||
+      !getModelLinkName(sdf, m_model_linkname)) {
+    ROS_ERROR("Failed to acquire regolith model SDF parameters");
+    return false;
+  }
+
+  // connect to all ROS services
+  ros::ServiceClient gz_get_phys_prop;
+  if (!createRosServiceClient<GetPhysicsProperties>(m_node_handle, 
+        SRV_GET_PHYS_PROPS, gz_get_phys_prop, false) ||
+      !createRosServiceClient<SpawnModel>(m_node_handle, 
+        SRV_SPAWN_MODEL, m_gz_spawn_model, true) ||
+      !createRosServiceClient<DeleteModel>(m_node_handle, 
+        SRV_DELETE_MODEL, m_gz_delete_model, true) ||
+      !createRosServiceClient<ApplyBodyWrench>(m_node_handle, 
+        SRV_APPLY_WRENCH, m_gz_apply_wrench, true) ||
+      !createRosServiceClient<BodyRequest>(m_node_handle, 
+        SRV_CLEAR_WRENCH, m_gz_clear_wrench, true)) {
+    ROS_ERROR("Failed to connect to all required ROS services");
+    return false;
+  }
```

Comment:
Hardly readable.. please extend the line limit to 120 chars.

Reply:
I agree it's hard to read, but because of bad indentation.  I suggest this, using the first clause as an example:
```
if (!createRosServiceClient<GetPhysicsProperties>(m_node_handle, 
                                                                                        SRV_GET_PHYS_PROPS,
                                                                                        gz_get_phys_prop, 
                                                                                        false)
```

which I think more readable to a wider audience than a really long line.